sql/hive/src/test/scala/org/apache/spark/sql/hive/CompressionCodecSuite.scala

```diff
@@ -0,0 +1,321 @@
+                sessionCompressionCodec, realCompressionCodec, tableSize)
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // When the amount of data is small, compressed data size may be larger than uncompressed one,
+  // so we just check the difference when compressionCodec is not NONE or UNCOMPRESSED.
+  private def checkTableSize(
+      format: String,
+      compressionCodec: String,
+      isPartitioned: Boolean,
+      convertMetastore: Boolean,
+      tableSize: Long): Boolean = {
+    format match {
+      case "parquet" =>
+        val uncompressedSize = if (!convertMetastore || isPartitioned) {
+          getUncompressedDataSizeByFormat(format, isPartitioned = true)
+        } else {
+          getUncompressedDataSizeByFormat(format, isPartitioned = false)
+        }
+
+        if (compressionCodec == "UNCOMPRESSED") {
+          tableSize == uncompressedSize
+        } else {
+          tableSize != uncompressedSize
+        }
+      case "orc" =>
+        val uncompressedSize = if (!convertMetastore || isPartitioned) {
+          getUncompressedDataSizeByFormat(format, isPartitioned = true)
+        } else {
+          getUncompressedDataSizeByFormat(format, isPartitioned = false)
+        }
+
+        if (compressionCodec == "NONE") {
+          tableSize == uncompressedSize
+        } else {
+          tableSize != uncompressedSize
+        }
+      case _ => false
+    }
+  }
+
+  def checkForTableWithCompressProp(format: String, compressCodecs: List[String]): Unit = {
+    Seq(true, false).foreach { isPartitioned =>
+      Seq(true, false).foreach { convertMetastore =>
+        checkTableCompressionCodecForCodecs(
+          format,
+          isPartitioned,
+          convertMetastore,
+          compressionCodecs = compressCodecs,
+          tableCompressionCodecs = compressCodecs) {
+          case (tableCompressionCodec, sessionCompressionCodec, realCompressionCodec, tableSize) =>
+            // For non-partitioned table and when convertMetastore is false, Expect session-level
+            // take effect, and in other cases expect table-level take effect
+            val expectCompressionCodec =
+              if (convertMetastore && !isPartitioned) sessionCompressionCodec
+              else tableCompressionCodec.get
+
+            assert(expectCompressionCodec == realCompressionCodec)
+            assert(checkTableSize(format, expectCompressionCodec,
+              isPartitioned, convertMetastore, tableSize))
+        }
+      }
+    }
+  }
+
+  def checkForTableWithoutCompressProp(format: String, compressCodecs: List[String]): Unit = {
+    Seq(true, false).foreach { isPartitioned =>
+      Seq(true, false).foreach { convertMetastore =>
+        checkTableCompressionCodecForCodecs(
+          format,
+          isPartitioned,
+          convertMetastore,
+          compressionCodecs = compressCodecs,
+          tableCompressionCodecs = List(null)) {
```

Comment:
`List(null)` -> `Nil`

Reply:
If change to `Nil`, the follow function may requires special handling of this situation. Set to `null` is to get a `None`.  Can we keep it?

``` scala
private def checkTableCompressionCodecForCodecs(
      format: String,
      isPartitioned: Boolean,
      convertMetastore: Boolean,
      compressionCodecs: List[String],
      tableCompressionCodecs: List[String])
      (assertionCompressionCodec: (Option[String], String, String, Long) => Unit): Unit = {
    withSQLConf(getConvertMetastoreConfName(format) -> convertMetastore.toString) {
      tableCompressionCodecs.foreach { tableCompression =>
        compressionCodecs.foreach { sessionCompressionCodec =>
          withSQLConf(getSparkCompressionConfName(format) -> sessionCompressionCodec) {
            // 'tableCompression = null' means no table-level compression
            val compression = Option(tableCompression)
            checkCompressionCodecForTable(format, isPartitioned, compression) {
              case (realCompressionCodec, tableSize) =>
                assertionCompressionCodec(
                  compression, sessionCompressionCodec, realCompressionCodec, tableSize)
            }
          }
        }
      }
    }
  }
```

Reply:
It does not work?

Reply:
If set to `Nil`,  ` foreach` statement will not go in and the testcase will do nothing. So a special handling must be done. Set to `null` will get a `None` value by ```val compression = Option(tableCompression)``` and it is exactly what I want.